fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,250 @@ impl ExecutorInner {
                 )
                 .await
                 .expect("Autocommit here can't fail");
-
-            None
-        } else {
-            Some(Box::pin(async move {
-                let (first_completed_result, _index, _unused_transitions) =
-                    select_all(transitions).await;
-                first_completed_result
-            }))
         }
+
+        transitions
     }
 
-    async fn execute_next_state_transitions(
+    async fn run_state_machines_executor_inner(
         &self,
-        global_context_gen: &ContextGen,
+        global_context_gen: ContextGen,
+        sm_update_rx: tokio::sync::mpsc::UnboundedReceiver<DynState>,
     ) -> anyhow::Result<()> {
+        // Re-add the same future  receiving on the channel.
+        // Not very elegant, but can't figure out how to make it a quine
+        // and don't know anything more clever.
+        fn re_add_sm_update_receiver(
+            futures: &mut FuturesUnordered<BoxFuture<'static, ExecutorLoopEvent>>,
+            mut sm_update_rx: mpsc::UnboundedReceiver<DynState>,
+        ) {
+            futures.push(Box::pin(async {
+                if let Some(new) = sm_update_rx.recv().await {
+                    ExecutorLoopEvent::New {
+                        state: new,
+                        sm_update_rx,
+                    }
+                } else {
+                    ExecutorLoopEvent::Disconnected
+                }
+            }));
+        }
+
         let active_states = self.get_active_states().await;
-        // TODO: use DB prefix subscription instead of polling
-        let mut active_state_count = active_states.len();
-        if active_states.is_empty() {
-            // FIXME: what to do in this case? Probably best to subscribe to DB eventually
-            trace!("No state transitions available, waiting before re-trying");
-            task::sleep(EXECUTOR_POLL_INTERVAL).await;
-            return Ok(());
+        trace!(target: LOG_CLIENT_REACTOR, "Starting active states: {:?}", active_states);
+        for (state, _meta) in active_states {
+            self.sm_update_tx
+                .send(state)
+                .expect("Must be able to send operation_id to own opened channel");
         }
-        trace!("Active states: {:?}", active_states);
 
-        let mut transitions = Vec::new();
-        for (state, meta) in active_states {
-            match self
-                .get_transition_for(state, meta, global_context_gen)
-                .await
-            {
-                None => {}
-                Some(t) => transitions.push(t),
-            }
+        /// All futures in the executor resolve to this type, so the handling
+        /// code can tell them apart.
+        enum ExecutorLoopEvent {
+            /// New active `DynState` showed up in the db and needs to be
```

Comment:
Fix: At this point we actually don't know if it's true, and only received a notification.